trunk/source/doc/perdb.1
.\"
.\" Source Version: 3.0
.\" Software Release #: LLNL-CODE-422942
.\"
.\" include cpyright.h
.\"

.TH PERDB 1 "02 September 2012"
.SH NAME
perdb \- persistent database
.SH SYNOPSIS
.B perdb [-c] [-d] [-f <db>] [-h] [-l] [-s] [<req>]

.SH DESCRIPTION

With POSIX systems the environment is a simple database
of the form VAR=VAL.  It is available to all programs run
on the system.  One of the features of the way the environment
is handled is that child processes inherit the environment
of their parents but cannot modify their parent's environment.

In some cases, such as configuring a code system for compilation,
a more symmetric database is desired.  Any process regardless
of relation to other processes can query, add to, or modify the
database.

Perdb was designed to work this way.  It is also designed to be
as transparent as possible.  See the examples given below.

Perdb works on a client/server model.  One can explicitly
launch a server using the -s option below.  Or a server will
be implicitly launched when perdb is run to perform any
other database operation.  The server can be explicitly shut
down with the quit: command or the server will save the database
in $PERDB.db and exit after the value of the environment
variable PERDB_IDLE_TIMEOUT passes without any transactions.

You can have the server and or the clients log transactions
by using the -l option.  The log file is $PERDB_PATH.log.


.SH OPTIONS

.TP
.B \-c
Create the database, removing any existing.
.TP
.B \-d
Do not run server as daemon (for debugging only).
.TP
.B \-f
Root path to database.
.TP
.B \-h
This help message.
.TP
.B \-l
Log transactions with the server.
.TP
.B \-s
Run as server.

.SH ENVIRONMENT VARIABLES

PERDB_PATH

This is the root path to a database.  It is not a directory
name, but a root name to which a suffix is attached for the
various files associated with a database.

PERDB_IDLE_TIMEOUT

This is the time that a server will run without processing
any transactions before it saves the database and exits.
The default is 60 seconds.

PERDB_IDLE_INTERVAL

This is the time that the server waits for input to become
available before checking on other aspects of its function.
The default is half of PERDB_IDLE_TIMEOUT.  You really do
not need to set this unless you are checking on performance
of the server.

.SH EXAMPLES

The following examples are taken from the perdb test suite.

.B Controlling the database

Here are several examples of selecting or controlling the
database.  In the first example the database is specified
on the command line.  In the second exampke, the database
is specified by the environment variable PERDB_PATH.  If
neither specification is used the default is ~/.perdb.

perdb -f ~/foo a=b

setenv PERDB_PATH ~/foo
perdb a=b

In these two cases the database would be ~/foo.db and
the log file would be ~/foo.log.


.B Set Operations

Here are several examples of setting variables.
In all of these it will be assumed that the environment
PERDB_PATH is set.

# set v1 - single token value

perdb v1=1

# set v2 - two token, single line list

perdb v2="baz blat"

# set v3 - angle bracket, multi-line list

perdb v3="< foo ;\e
.BR
 bar >"

# set v4 = double quoted, parenthesized, multi-line list

perdb v4="( foo ;\e
.BR
 bar )"

# set v5 = single quoted, parenthesized, multi-line list

perdb v5='( foo ;\e
.BR
 bar )'

# set v6 = single quoted, curly bracket, multi-line list

perdb v6='{ foo ;\e
.BR
 bar }'

# set v7 = single quoted, double quoted, multi-line list

perdb v7='" foo ;\e
.BR
 bar "'


.B Get Operations

These are examples of getting variables from a database.
In all of these it will be assumed that the environment
PERDB_PATH is set.

# get v1

perdb v1

.B Def Operations

These are examples of querying a database for existence
of a variable:

perdb v1 \e?

.B Conditional Operations

These are examples of setting the values of variables if
and only if they have no current value:

perdb v2 =\e? foo

.B Perdb Commands

In all of these it will be assumed that the environment
PERDB_PATH is set.

# write all variables to file $PERDB_PATH.db

perdb save:

# write variable v1 to file $PERDB_PATH.db

perdb save: v1

# write all variable v1 to file $PERDB_PATH.dba.db

perdb save dba: v1

# read all variables from file $PERDB_PATH.db into the database

perdb load:

# read all variables from file $PERDB_PATH.dba.db into the database

perdb load dba:

# reset the database, removing all variables

perdb reset:

# tell the server to exit

perdb quit:

.SH AMENITIES FOR CSH

If you include the PACT file functions-pact.csh in your CSH scripts
you will have the following aliases which make using perdb
more convenient in your shell scripts:

dbset <var> <val>

Set variable <var> to <val> in the database and the
current environment.  Use dbset instead of setenv.  This
also matches the C call dbset.

dbget <var>

Import the database variable <var> into the current environment.
Use dbget before referencing the variable.  This also matches
the C call dbget.

dbdef <var>

Query the database for existence of variable <var>.  This also
matches the C call dbdef.  Check $status for the result.

dbini <var> <val>

Initialize variable <var> to <val> in the database if and only
if <var> does not already have a value.  This also matches the
C call dbinitv.

dbexp <var>

Export the current value of environment variable <var>
back to the database.  This also matches the C call dbexp.


.SH BUGS
 
Not optimized for performance.  Also transactions sometimes
lost on highly loaded systems.



source/psh/perdb.c
/*
 * PERDB.C - implement persistent database access
 *
 * Source Version: 3.0
 * Software Release #: LLNL-CODE-422942
 *
 * #include "cpyright.h"
 *
 */

#include "common.h"
#include "libpsh.c"
#include "libdb.c"

static database
 *db = NULL;

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SIGDONE - handle signals meant to end session */

static void sigdone(int sig)
   {

    if (db != NULL)
       {log_activity(db->flog, dbg_db, 4,
		     "SERVER", "signalled %d - done", sig);
	close_sock(db->root);
	db_srv_save(-1, db);
	unlink(db->fpid);};

    exit(sig);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SIGRESTART - handle signals meant to restart session */

static void sigrestart(int sig)
   {

    db_srv_restart(db);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SRV_SAVE_DB - save the database */

static char *srv_save_db(client *cl, char *fname, char **var, const char *fmt)
   {int fd, ok;
    char s[MAXLINE];
    FILE *fp;
    database *db;
    static char t[MAXLINE];

    fd = cl->fd;
    db = cl->db;

/* figure out where results go */
    if ((fname == NULL) || (strcmp(fname, "stdout") == 0))
       fp = NULL;
    else
       {if (fname[0] == '/')
	   nstrncpy(s, MAXLINE, fname, -1);
        else
	   snprintf(s, MAXLINE, "%s.%s.db", db->root, fname);

	fp = fopen(s, "w");
	if (fp == NULL)
	   {snprintf(t, MAXLINE, "could not open %s - save %s",
		     s, strerror(errno));
	    return(t);};
        fname = path_tail(s);};
	   
/* do the work */
    ok = save_db(fd, db, var, fp, fmt);
    ASSERT(ok == 0);

    if (fp != NULL)
       fclose(fp);
	
/* make up the ending message */
    if ((IS_NULL(fmt) == FALSE) &&
	((strncmp(fmt, "setenv ", 7) == 0) ||
	 (strncmp(fmt, "export ", 7) == 0)))
       snprintf(t, MAXLINE, "%s", cwhich("true"));

     else if (fname == NULL)
       {if (var == NULL)
	   snprintf(t, MAXLINE, "saved database");
        else
	   snprintf(t, MAXLINE, "saved variables");}

    else
       {if (var == NULL)
	   snprintf(t, MAXLINE, "saved database to %s",
		    fname);
	else
	   snprintf(t, MAXLINE, "saved variables to %s",
		    fname);};

    return(t);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SRV_LOAD_DB - load the database */

static char *srv_load_db(client *cl, char *fname, char *var)
   {char s[MAXLINE];
    char *root;
    FILE *fp;
    database *db;
    static char t[MAXLINE];

    db   = cl->db;
    root = cl->root;

    if (fname == NULL)
       fp = NULL;

    else
       {if (file_exists(fname) == TRUE)
	   nstrncpy(s, MAXLINE, fname, -1);
        else
	   snprintf(s, MAXLINE, "%s.%s.db", root, fname);

	fp = fopen(s, "r");
        if (fp == NULL)
	   {snprintf(t, MAXLINE, "could not open %s - load %s",
		     s, strerror(errno));
	    return(t);};
        fname = path_tail(s);};
	   
    load_db(db, var, fp);

    if (fname == NULL)
       {if (var == NULL)
	   snprintf(t, MAXLINE, "loaded database");
        else
	   snprintf(t, MAXLINE, "loaded %s", var);}

    else
       {if (var == NULL)
	   snprintf(t, MAXLINE, "loaded database from %s",
		    fname);
	else
	   snprintf(t, MAXLINE, "loaded %s from %s",
		    var, fname);};

    if (fp != NULL)
       fclose(fp);
	
    return(t);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* CHECK_FD - verify/log fd set in SRV */

static void check_fd(connection *srv, char *flog)
   {int fd;
    char s[MAXLINE];
    fd_set rfs;

    rfs = srv->afs;

    s[0] = '\0';
    for (fd = 0; fd < FD_SETSIZE; ++fd)
        {if (FD_ISSET(fd, &rfs))
            vstrcat(s, MAXLINE, " %d", fd);};

    log_activity(flog, dbg_db, 4, "SERVER", "monitoring: %s", s);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* ADD_FD - add FD to the set in SRV */

static void add_fd(connection *srv, char *flog, int fd)
   {

    FD_SET(fd, &srv->afs);
    log_activity(flog, dbg_db, 1, "SERVER", "add fd %d", fd);

    check_fd(srv, flog);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* REMOVE_FD - add FD to the set in SRV */

static void remove_fd(connection *srv, char *flog, int fd)
   {

    FD_CLR(fd, &srv->afs);
    log_activity(flog, dbg_db, 1, "SERVER", "remove fd %d", fd);

    check_fd(srv, flog);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* NEW_CONNECTION - make a new connection on SFD and add it to AFS */

static void new_connection(char *root, connection *srv)
   {int fd;
    char *flog;
    socklen_t sz;

    flog = name_log(root);

    sz = sizeof(srv->sck);

    fd = accept(srv->server, (struct sockaddr *) &srv->sck, &sz);
    if (fd > 0)
       add_fd(srv, flog, fd);
    else
       {close(fd);
	log_activity(flog, dbg_db, 1, "SERVER",
		     "accept error - %s (%d)",
		     strerror(errno), errno);};

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* TERM_CONNECTION - finish off client connection CL */

static void term_connection(client *cl)
   {int fd;
    char *root, *flog;
    connection *srv;

    fd   = cl->fd;
    root = cl->root;
    srv  = cl->server;
    flog = name_log(root);

    remove_fd(srv, flog, fd);

    cl->fd = connect_close(fd, cl, NULL);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_SAVE - save the database variables
 *         - syntax is
 *         -    save:[<name>] <var>*
 *         - where
 *         -   <name> is a database name
 *         -          if <name> is stdout dumps to stdout
 *         -          else if <name> is full path use it
 *         -          otherwise database will be named <root>.<name>.db
 *         -          default is <root>.db
 *         -   <var>  variable name to be saved
 *         -          if no variables specified all are saved
 */

static char *do_save(client *cl, char *s, const char *fmt)
   {char **var, *val, *fname, *p;

    p = s;
    if ((*p == '\0') || (strchr(" \t\n\f", *p) != NULL))
       fname = NULL;
    else
       {fname = p;

	p = strpbrk(fname, " \t\n\f");
        if (p != NULL)
	   *p++ = '\0';};

    var = tokenize(p, " \t\n\f");
    if ((var != NULL) && (var[0] == NULL))
       {free_strings(var);
        var = NULL;};

    val = srv_save_db(cl, fname, var, fmt);

    free_strings(var);

    return(val);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_LOAD - load the database variables
 *         - syntax is
 *         -    load:[<name>] <var>*
 *         - where
 *         -   <name> is a database name
 *         -          if <name> is stdout dumps to stdout
 *         -          else if <name> is full path use it
 *         -          otherwise database will be named <root>.<name>.db
 *         -          default is <root>.db
 *         -   <var>  variable name to be loaded
 *         -          if no variables specified all are loaded
 */

static char *do_load(client *cl, char *s)
   {char *var, *val, *fname, *p;

    p = s + 5;
    if ((*p == '\0') || (strchr(" \t\n\f", *p) != NULL))
       fname = NULL;
    else
       {fname = p;

	p = strpbrk(fname, " \t\n\f");
	if (p != NULL)
	   *p++ = '\0';};

    key_val(&var, NULL, p, " \t\n");

    val = srv_load_db(cl, fname, var);

    return(val);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_COND_INIT - handle conditional initialization */

static char *do_cond_init(database *db, char *s)
   {int st;
    char *var, *val, *nvl;

    key_val(&var, &nvl, s, "=? \t\n");

    val = get_db(db, var);

    st = (val != cnoval());
    if (st == FALSE)
       val = put_db(db, var, nvl);

    return(val);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_DEFD - handle defined query */

static char *do_defd(database *db, char *s)
   {int st;
    char *var, *val, *nvl;

    key_val(&var, &nvl, s, "? \t\n");

    val = get_db(db, var);

    st = (val != cnoval());
    if (IS_NULL(nvl) == TRUE)
       {if (st == TRUE)
	   val = "defined{TRUE}";
        else
	   val = "defined{FALSE}";}

    else if (st == FALSE)
       val = put_db(db, var, nvl);

    return(val);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* DO_SET_GET - handle variable set or get operations */

static char *do_set_get(database *db, char *s)
   {char *var, *val;
    static char t[MAXLINE];

    key_val(&var, &val, s, "= \t\n");

    if (val == NULL)
       val = get_db(db, var);

    else
       {val = trim(val, BACK, " \t");
	val = put_db(db, var, val);};

    if ((val != NULL) &&
	(strchr("'\"(", val[0]) == NULL) && 
	(strpbrk(val, " \t") != NULL))
       {snprintf(t, MAXLINE, "\"%s\"", val);
	val = t;};

    return(val);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* PROC_CONNECTION - process input on client connection CL
 *                 - return -1 on error
 *                 -         0 on end of connection
 *                 -         1 if otherwise successful
 */

static int proc_connection(client *cl)
   {int rv, nb, to;
    char s[MAXLINE];
    char *val, *fmt;
    database *db;

    db = cl->db;

    to = 60;
    rv = 1;

    memset(s, 0, MAXLINE);

    nb = comm_read(cl, s, MAXLINE, to);

/* decide on the action to take */
    if (nb <= 0)
       rv = -1;

    else

/* quit session */
       {if (strncmp(s, "quit:", 5) == 0)
	   {rv  = 0;
	    val = "done";}

/* client is exiting */
	else if (strncmp(s, "fin:", 4) == 0)
	   {term_connection(cl);
	    val = NULL;}

/* reset database */
        else if (strncmp(s, "reset:", 6) == 0)
	   {reset_db(db);
	    val = "reset";}

/* import variables to CSH environment */
	else if (strncmp(s, "csh:", 4) == 0)
	   {fmt = "setenv %s %s ;";
	    val = do_save(cl, s+4, fmt);}

/* import variables to SH environment */
	else if (strncmp(s, "sh:", 3) == 0)
	   {fmt = "export %s=%s ;";
	    val = do_save(cl, s+3, fmt);}

/* save database */
	else if (strncmp(s, "save:", 5) == 0)
	   val = do_save(cl, s+5, "%s=%s");

/* load database */
	else if (strncmp(s, "load:", 5) == 0)
           val = do_load(cl, s);

/* variable conditional init */
	else if (strstr(s, "=\?") != NULL)
	   val = do_cond_init(db, s);

/* variable defined? */
	else if (strchr(s, '?') != NULL)
           val = do_defd(db, s);

/* variable set/get */
	else
	   val = do_set_get(db, s);

	if (val != NULL)
	   {nb = comm_write(cl, val, 0, 10);
	    nb = comm_write(cl, EOM, 0, 10);};};

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* ASYNC_SERVER - run a fully asynchronous server */

static void async_server(client *cl)
   {int fd, nr, ok;
    char *root, *flog;
    fd_set rfs;
    connection *srv;

    root = cl->root;
    srv  = cl->server;

    flog = name_log(root);

    if (listen(srv->server, 5) >= 0)
       {int ng, nb, nbmax, dt, tmax;
	char *s;
        struct timeval t;

/* set idle timeout and idle interval
 * if no input comes in within the idle timeout window stop and return
 * have the select timeout at the idle interval
 */
        s = cgetenv(FALSE, "PERDB_IDLE_TIMEOUT");
	if (IS_NULL(s) == TRUE)
	   tmax = 60;
	else
	   {tmax = atoi(s);
	    log_activity(flog, dbg_db, 4, "SERVER",
			 "PERDB_IDLE_TIMEOUT = %d", tmax);};

        s = cgetenv(FALSE, "PERDB_IDLE_INTERVAL");
	if (IS_NULL(s) == TRUE)
	   dt = tmax >> 1;
	else
	   {dt = atoi(s);
	    log_activity(flog, dbg_db, 4, "SERVER",
			 "PERDB_IDLE_INTERVAL = %d", dt);};

/* maximum number of consecutive 0 length reads
 * indicating dropped connection
 */
	nb    = 0;
	nbmax = 10;

	ng = 0;
	for (ok = TRUE; (ok == TRUE) && (ng < tmax) && (nb < nbmax); )
	    {check_fd(srv, flog);

/* timeout the select every DT seconds */
	     t.tv_sec  = dt;
	     t.tv_usec = 0;

	     rfs = srv->afs;
	     nr  = select(FD_SETSIZE, &rfs, NULL, NULL, &t);
	     if (nr > 0)
	        {for (fd = 0; (fd < FD_SETSIZE) && (ok == TRUE); ++fd)
		     {if (FD_ISSET(fd, &rfs))
			 {log_activity(flog, dbg_db, 4, "SERVER",
				       "data available on %d (server %d)",
				       fd, srv->server);

/* new connection request */
			  if (fd == srv->server)
			     new_connection(root, srv);

/* data arriving on an existing connection */
			  else
			     {cl->fd = fd;

			      ok = proc_connection(cl);
                              if (ok == -1)
				 {ok = TRUE;
				  remove_fd(srv, flog, fd);
				  nb++;}
			      else
				 nb = 0;

			      sched_yield();};};};
		  ng = 0;}
	     else
	        ng += dt;};

	if (ok != TRUE)
	   log_activity(flog, dbg_db, 4, "SERVER", "done by command");
	else if (ng >= tmax)
	   log_activity(flog, dbg_db, 4, "SERVER",
			"done by time: %d >= %d", ng, tmax);
	else if (nb >= nbmax)
	   log_activity(flog, dbg_db, 4, "SERVER",
			"done by failed reads: %d >= %d", nb, nbmax);}

    else
       log_activity(flog, dbg_db, 1, "SERVER",
		    "async_server error (%s - %d)",
		    strerror(errno), errno);

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* SERVER - run the server side of the database */

static int server(char *root, int init, int dmn)
   {char *flog;
    client *cl;

    if ((dmn == FALSE) || (demonize() == TRUE))
       {signal(SIGTERM, sigdone);
	signal(SIGINT, sigdone);
	signal(SIGHUP, sigrestart);

	flog = name_log(root);

	log_activity(flog, dbg_db, 1, "SERVER", "start server");

	cl = make_client(root, SERVER);

	cl->db = db_srv_open(root, init);
	if (cl->db != NULL)
	   {db = cl->db;

	    async_server(cl);

	    db_srv_save(-1, cl->db);
	    db_srv_close(cl->db);};

	free_client(cl);

	log_activity(flog, dbg_db, 1, "SERVER", "end server");};

    return(0);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* EXCHANGE - do a transaction from the client side of the database
 *          - if LTR is TRUE strip off the quotes the server side added
 */

static int exchange(char *root, int ltr, char *req)
   {int i, n, rv;
    char **ta;
    client *cl;

    cl = make_client(root, CLIENT);
    ta = _db_clnt_ex(cl, TRUE, req);
    free_client(cl);

    rv = (ta != NULL);

/* print the reply */
    n = lst_length(ta);
    for (i = 0; i < n; i++)
        {if (strcmp(ta[i], "defined{TRUE}") == 0)
	    rv = TRUE;
	 else if (strcmp(ta[i], "defined{FALSE}") == 0)
	    rv = FALSE;
	 else if (ltr == TRUE)
	    printf("%s\n", strip_quote(ta[i]));
	 else
	    printf("%s\n", ta[i]);};

    lst_free(ta);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

/* HELP - print usage info */

static void help(void)
   {
    printf("\n");
    printf("Usage: perdb [-c] [-d] [-e] [-f <db>] [-h] [-l] [-s] [<req>]\n");
    printf("   c   create the database, removing any existing\n");
    printf("   d   do not run server as daemon\n");
    printf("   e   exact - do not strip off quotes\n");
    printf("   f   root path to database\n");
    printf("   h   this help message\n");
    printf("   l   log transactions with the server\n");
    printf("   s   run as server\n");
    printf("\n");

    return;}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/

int main(int c, char **v)
   {int i, rv, srv, dmn, init, ltr;
    char root[MAXLINE], r[MAXLINE], req[MAXLINE];
    char *except[] = {"PATH", "PERDB_IDLE_TIMEOUT",
		      "PERDB_IDLE_INTERVAL", "PERDB_PATH",
		      NULL};

    r[0] = '\0';
    srv  = FALSE;
    init = FALSE;
    dmn  = TRUE;
    ltr  = FALSE;

/* everything up to the first token not beginning with '-' is a flag */
    for (i = 1; i < c; i++)
        {if (v[i][0] == '-')
	    {switch (v[i][1])
                {case 'c' :
                      init = TRUE;
		      break;
                 case 'd' :
                      dmn = FALSE;
		      break;
                 case 'e' :
                      ltr = TRUE;
		      break;
                 case 'f' :
                      nstrncpy(r, MAXLINE, v[++i], -1);
		      break;
		 case 'h' :
                      help();
                      return(1);
		 case 'l' :
                      dbg_sock = TRUE;
                      dbg_db   = TRUE;
                      break;
		 case 's' :
                      srv = TRUE;
                      break;};}
	 else
	    break;};

/* everything from here is the request */
    req[0] = '\0';
    for (; i < c; i++)
        {nstrcat(req, MAXLINE, v[i]);
	 nstrcat(req, MAXLINE, " ");};

    if (IS_NULL(r) == TRUE)
       {if (cdefenv("PERDB_PATH") == TRUE)
           nstrncpy(r, MAXLINE, cgetenv(TRUE, "PERDB_PATH"), -1);};

    if (IS_NULL(r) == TRUE)
       snprintf(r, MAXLINE, "%s/.perdb", cgetenv(TRUE, "HOME"));

    full_path(root, MAXLINE, NULL, r);

    rv = cclearenv(except);

    rv = TRUE;
    if (srv == TRUE)
       rv = server(root, init, dmn);

    else if (IS_NULL(req) == FALSE)
      {LAST_CHAR(req) = '\0';
       rv = exchange(root, ltr, req);};

    rv = (rv == 0);

    return(rv);}

/*--------------------------------------------------------------------------*/
/*--------------------------------------------------------------------------*/
